### PR TITLE
Bug fixes Campaign builder

### PR DESCRIPTION
**Pull request by Alexandr57**
Fixed most bugs Campaign builder

### CODE REVIEW

File: Utils/Campaign builder/Unit1.pas [hunk 1]
```diff
@@ -251,6 +255,12 @@ procedure TForm1.btnLoadCMPClick(Sender: TObject);
   fSelectedMap := -1;
   fSelectedNode := -1;
 
+  //From UpdateList ruled here since caused a bug with entering
+  //More details about the bug:
+  {http://gifok.net/images/2017/02/14/2017-02-14_16-01-27.gif}
```
Comment (lewinjh) [inline Utils/Campaign builder/Unit1.pas]: Please don't write descriptions of your changes and links to gifs in the code. Post them in the description of the pull request.

In this case I would say this change doesn't require a comment in the code (instead post the gif and reasoning for your changes in the pull request description).

File: Utils/Campaign builder/Unit1.pas [hunk 2]
```diff
@@ -302,6 +312,12 @@ procedure TForm1.seMapCountChange(Sender: TObject);
 begin
   if fUpdating then Exit;
 
+  //Bug fixes flaws SpinEdit component which makes checking the entered number after the OnChange event
+  //If you remove the "if (seMapCount.Value > MAX_CAMP_MAPS)or(seMapCount.Value < seMapCount.MinValue) then" there is a problem shown here:
+  //http://gifok.net/images/2017/02/14/2017-02-14_16-01-27.gif
```
Comment (lewinjh) [inline Utils/Campaign builder/Unit1.pas]: As I said above, this comment should be in the pull request description not the code.

Here you should write a comment like "//SpinEdit can call OnChange with a value outside of the valid range, so validate it first"

File: Utils/Campaign builder/Unit1.pas [hunk 3]
```diff
@@ -302,6 +312,12 @@ procedure TForm1.seMapCountChange(Sender: TObject);
 begin
   if fUpdating then Exit;
 
+  //Bug fixes flaws SpinEdit component which makes checking the entered number after the OnChange event
+  //If you remove the "if (seMapCount.Value > MAX_CAMP_MAPS)or(seMapCount.Value < seMapCount.MinValue) then" there is a problem shown here:
+  //http://gifok.net/images/2017/02/14/2017-02-14_16-01-27.gif
+  if (seMapCount.Value > MAX_CAMP_MAPS)or(seMapCount.Value < seMapCount.MinValue) then
+    seMapCount.Value := EnsureRange(seMapCount.Value, seMapCount.MinValue, MAX_CAMP_MAPS);
```
Comment (lewinjh) [inline Utils/Campaign builder/Unit1.pas]: There is no need for the if statement, you can always run EnsureRange here (result will be the same).

File: Utils/Campaign builder/Unit1.pas [hunk 4]
```diff
@@ -317,6 +333,12 @@ procedure TForm1.MapChange(Sender: TObject);
 begin
   if fUpdating or (fSelectedMap = -1) then Exit;
 
+  //Bug fixes flaws SpinEdit component which makes checking the entered number after the OnChange event
+  //If you remove the "if (seNodeCount.Value > MAX_CAMP_NODES)or(seNodeCount.Value < seNodeCount.MinValue) then" there is a problem shown here:
+  //http://gifok.net/images/2017/02/14/2017-02-14_16-01-27.gif
+  if (seNodeCount.Value > MAX_CAMP_NODES)or(seNodeCount.Value < seNodeCount.MinValue) then
+    seNodeCount.Value := EnsureRange(seNodeCount.Value, seNodeCount.MinValue, MAX_CAMP_NODES);
```
Comment (lewinjh) [inline Utils/Campaign builder/Unit1.pas]: Same as my two comments on your other SpinEdit change

**Review by lewinjh — CHANGES_REQUESTED** [review]

**Comment by Alexandr57** [comment]
@lewinjh 
If you remove this code:
if (seMapCount.Value > MAX_CAMP_MAPS)or(seMapCount.Value < seMapCount.MinValue) then
That there is such a problem:
http://gifok.net/images/2017/02/14/2017-02-14_16-01-27.gif
This is due to the fact that:
seMapCount.Value: = EnsureRange (seMapCount.Value, seMapCount.MinValue, MAX_CAMP_MAPS);
Re raises the event
OnChange
What do you advise?

**Comment by Alexandr57** [comment]
> Comments Rus
```
Вот так тебе Kromster80 не нравится:
procedure TSpinEdit.CheckValue;
begin
  if not inRange(Value, MinValue, MaxValue) then
  Value := EnsureRange(Value, MinValue, MaxValue);
end;

А вот так:

procedure TSpinEdit.CheckValue;
begin
  Value := EnsureRange(Value, MinValue, MaxValue);
end;

Вызывает баг делая невозможным ввод вручную
```

> Comments Eng

```
So you do not like Kromster80:

procedure TSpinEdit.CheckValue;
begin
  if not inRange(Value, MinValue, MaxValue) then
  Value := EnsureRange(Value, MinValue, MaxValue);
end;

And like this:

procedure TSpinEdit.CheckValue;
begin
  Value := EnsureRange(Value, MinValue, MaxValue);
end;

It is a bug making it impossible to input manually
```

> Here is the result of such an option code:
```
procedure TForm1.seMapCountChange(Sender: TObject);
begin
  if fUpdating then Exit;

  seMapCount.CheckValue;

  C.MapCount := seMapCount.Value;

  if fSelectedMap > C.MapCount - 1 then
    fSelectedMap := -1;

  UpdateList;
  UpdateFlagCount;
  RefreshFlags;
end;
procedure TForm1.MapChange(Sender: TObject);
begin
  if fUpdating or (fSelectedMap = -1) then Exit;

  seNodeCount.CheckValue;

  C.Maps[fSelectedMap].NodeCount := seNodeCount.Value;

  if fSelectedNode > C.Maps[fSelectedMap].NodeCount - 1 then
    fSelectedNode := -1;

  UpdateList;
  UpdateNodeCount;
  RefreshFlags;
end;
...
procedure TSpinEdit.CheckValue;
begin
  Value := EnsureRange(Value, MinValue, MaxValue);
end;
```
Result:

PS: I'm in SE introduced the 1 digit. He substituted a second.

**Comment by lewinjh** [comment]
Don't set the spin edit value from within the change handler. Instead make a local copy of the value which has been checked with EnsureRange and use that within the change handler. We shouldn't be setting the value back to the spin edit, just validating it before we use it.

**Comment by Alexandr57** [comment]
@lewinjh 
That maximum value is set at the beginning:
```
procedure TForm1.FormCreate(Sender: TObject);
begin
  DoubleBuffered := True; //Makes images drag around smoothly
  ScrollBox1.DoubleBuffered := True;

  Caption := 'Campaign Builder (' + GAME_REVISION + ')';

  C := TKMCampaign.Create;
  fSelectedMap := -1;

  //This line corrects a bug in UpdateList namely C.CampName line that returns at the start of the program #0#0#0
  edtShortNameChange(nil);

  seMapCount.MaxValue := MAX_CAMP_MAPS;
  seNodeCount.MaxValue := MAX_CAMP_NODES;

  fSprites := TKMSpritePackEdit.Create(rxCustom, nil);

  seMapCountChange(nil); //Initialise it to 1 map
end;
```
If nothing is done then that's what will

Or do you propose to write a new event that will occur before the OnChange

**Comment by Alexandr57** [comment]
image
const
  MAX_CAMP_MAPS = 64;
  MAX_CAMP_NODES = 64;

**Comment by lewinjh** [comment]
I've committed my proposed solution, take a look

**Comment by Alexandr57** [comment]
@lewinjh 
Sorry. Only now realized!
It can be like this:
```
C.MapCount := seMapCount.GetValue;
C.Maps[fSelectedMap].NodeCount := seNodeCount.GetValue;
...
function TSpinEdit.GetValue: Integer;
begin
  Result := EnsureRange(Value, MinValue, MaxValue);
end;
```
It'll do?

**Comment by lewinjh** [comment]
No I don't like that. We don't need to customise the spin edit class. Just do the EnsureRange when we take the value from the view (spin edit) to the model (campaign data). Look at what I committed in master.

**Comment by Alexandr57** [comment]
Ok

**Review by lewinjh — APPROVED** [review]

**Pull request merged**